### PR TITLE
Executor service merge

### PR DESCRIPTION
**Pull request by ylil93**

### CODE REVIEW

File: sdk/cpp/core/docsgen/api/services/executor_service.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+.. cpp:namespace:: ydk
+
+.. cpp:class:: ydk::ExecutorService : public Service
+
+Executor Service class for supporting execution of RPCs.
+
+    .. cpp:function:: ExecutorService()
+
+    .. cpp:function:: bool create(path::ServiceProvider & provider, Entity & entity)
```
Comment (abhikeshav) [inline sdk/cpp/core/docsgen/api/services/executor_service.rst]: Why is this line here?

File: sdk/cpp/core/docsgen/api/services/executor_service.rst [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    .. cpp:function:: std::shared_ptr<Entity> execute_rpc(NetconfServiceProvider & provider, Entity & entity)
```
Comment (abhikeshav) [inline sdk/cpp/core/docsgen/api/services/executor_service.rst]: `Entity & entity` should be `Entity & rpc`

File: sdk/cpp/core/docsgen/api/services/executor_service.rst [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+        Create the entity.
+        
+        :param provider: An instance of :cpp:class:`NetconfServiceProvider<ydk::NetconfServiceProvider>`
+        :param entity: An instance of :cpp:class:`Entity<ydk::Entity>` class defined under a bundle
```
Comment (abhikeshav) [inline sdk/cpp/core/docsgen/api/services/executor_service.rst]: How will anyone reading this know this is supposed to be an rpc object?

File: sdk/cpp/core/src/executor_service.cpp [hunk 4]
```diff
@@ -35,59 +35,124 @@ namespace ydk{
...
-    // // Get the operation
-    // auto operation = entity.get_entity_path(entity.parent).path;
-
-    // // Set the data tag
-    // string data_tag = "entity";
-
-    // // Set the config flag
-    // bool set_config_flag = false;
+    // Get the operation - RPC Name
+    // auto operation = rpc_entity.get_entity_path(rpc_entity.parent).path;
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Remove commented line

File: sdk/cpp/core/src/executor_service.cpp [hunk 5]
```diff
@@ -35,59 +35,124 @@ namespace ydk{
...
+    }
+    else
+        return nullptr;
+}
 
-    // // Get the root schema node
-    // path::RootSchemaNode* root_schema = provider.get_root_schema();
-    // unique_ptr<ydk::path::Rpc> ydk_rpc { root_schema->rpc(operation) };
-    // // to do: implement way to walk through children to get what to create
-    // ydk_rpc->input()->create("source/candidate");    // -- hard coding
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Why are these comments here?

File: sdk/cpp/core/src/executor_service.cpp [hunk 6]
```diff
@@ -35,59 +35,124 @@ namespace ydk{
...
+
+        create_from_children(children, rpc_input, root_schema);
+    }
+}
 
-    // // Create netconf payload
-    // path::CodecService codec_service{};
-    // std::string netconf_payload{"<rpc xmlns=\"urn:ietf:params:xml:ns:netconf:base:1.0\">"};
-    // netconf_payload+=codec_service.encode(ydk_rpc->input(), path::CodecService::Format::XML, false);
-    // netconf_payload+="</rpc>";
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Why are these commented lines here?

File: sdk/cpp/core/src/executor_service.cpp [hunk 7]
```diff
@@ -35,59 +35,124 @@ namespace ydk{
...
+        std::string temp_path = "";
+        if (path != "")
+            temp_path = path + '/';
+        temp_path = temp_path + child.first;
+        rpc_input.create(temp_path, child.second.value);
+    }
+}
 
-    // // Execute payload -- removed... rewrite this code!!!
-    // std::string reply = provider.execute_payload(netconf_payload);
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Commented line

File: sdk/cpp/core/src/executor_service.cpp [hunk 8]
```diff
@@ -35,59 +35,124 @@ namespace ydk{
...
+            YLOG_DEBUG("Creating child '{}': {}",child.first, child.second->get_entity_path(child.second->parent).path);
+
+            std::string payload = get_data_payload(*child.second, root_schema);
+            rpc_input.create(child.first, payload);
+        }
+    }
+}
 
-    // // Scan reply for OK
-    // return reply.find("<ok/>") != string::npos;
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Commented line

File: sdk/cpp/core/src/executor_service.cpp [hunk 9]
```diff
@@ -35,59 +35,124 @@ namespace ydk{
...
-    //     throw YCPPIllegalStateError{"Failed to create data tree"};
-    // }
-
-    // std::string payload{"<rpc xmlns=\"urn:ietf:params:xml:ns:netconf:base:1.0\">"};
-    // payload+=codec_service.encode(input, path::CodecService::Format::XML, false);
-    // payload+="</rpc>";
-    // BOOST_LOG_TRIVIAL(debug) <<"=============Generating payload=============";
-    // BOOST_LOG_TRIVIAL(debug) <<payload;
-    // BOOST_LOG_TRIVIAL(debug) <<"=========================="<<endl;
-    // return payload;
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Commented lines

File: sdk/cpp/tests/main.cpp.in [hunk 10]
```diff
@@ -10,7 +10,7 @@ int main( int argc, char* argv[] )
 {
   // global setup...
   auto logger = spdlog::stdout_color_mt("ydk");
-  logger->set_level(spdlog::level::err);
+  logger->set_level(spdlog::level::debug);
```
Comment (abhikeshav) [inline sdk/cpp/tests/main.cpp.in]: This should not be changed

File: sdk/cpp/tests/test_executor_service.cpp [hunk 11]
```diff
@@ -28,41 +30,207 @@
...
+    ExecutorService es{};
+
+    ydk::ietf_netconf::DiscardChangesRpc rpc{};
+
+    std::shared_ptr<Entity> reply = es.execute_rpc(provider, rpc);
+    bool result = reply == nullptr;
+    REQUIRE(result);
 }
 
-// TEST_CASE("execute_validate_rpc_source_1")
```
Comment (abhikeshav) [inline sdk/cpp/tests/test_executor_service.cpp]: This needs to be fixed or removed

File: sdk/cpp/tests/test_executor_service.cpp [hunk 12]
```diff
@@ -92,7 +260,7 @@ TEST_CASE("dummy")
 //     REQUIRE(reply);
 // }
 
-// TEST_CASE("execute_validate_rpc_source_5")
+// TEST_CASE("es_validate_rpc_5")
 // {
 //     ydk::path::Repository repo{TEST_HOME};
 //     NetconfServiceProvider provider{repo, "127.0.0.1", "admin", "admin", 12022};
```
Comment (abhikeshav) [inline sdk/cpp/tests/test_executor_service.cpp]: Commented lines

File: sdk/cpp/tests/main.cpp.in (hunk 10 above)
Comment (psykokwak4) [inline sdk/cpp/tests/main.cpp.in]: Setting to `debug` explodes Travis CI log, add will cause build failure if it exceed 4 MB limit.

File: sdk/cpp/core/src/executor_service.cpp [hunk 13]
```diff
@@ -33,61 +33,126 @@ using namespace std;
 
 namespace ydk{
 
-// static string get_data_payload(Entity & entity, string data_tag, path::RootSchemaNode & root_schema);
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Remove commented lines

File: sdk/cpp/core/src/executor_service.cpp [hunk 14]
```diff
@@ -33,61 +33,126 @@ using namespace std;
...
 }
 
-bool ExecutorService::execute_rpc(NetconfServiceProvider & provider, Entity & entity)
+shared_ptr<Entity> ExecutorService::execute_rpc(NetconfServiceProvider & provider,
+    Entity & rpc_entity, std::shared_ptr<Entity> top_entity)
 {
-    // // Get the operation
-    // auto operation = entity.get_entity_path(entity.parent).path;
+    // Get the operation - RPC Name
+    // auto operation = rpc_entity.get_entity_path(rpc_entity.parent).path;
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Remove commented lines

File: sdk/cpp/core/src/executor_service.cpp [hunk 15]
```diff
@@ -33,61 +33,126 @@ using namespace std;
...
-    // // Set the config flag
-    // bool set_config_flag = false;
+        if (entity_path.path != "input")
+            path = path + entity_path.path;
 
-    // // Get the root schema node
-    // path::RootSchemaNode* root_schema = provider.get_root_schema();
-    // unique_ptr<ydk::path::Rpc> ydk_rpc { root_schema->rpc(operation) };
-    // // to do: implement way to walk through children to get what to create
-    // ydk_rpc->input()->create("source/candidate");    // -- hard coding
```
Comment (abhikeshav) [inline sdk/cpp/core/src/executor_service.cpp]: Remove commented lines

File: sdk/cpp/tests/test_executor_service.cpp [hunk 16]
```diff
@@ -28,80 +33,318 @@
...
-// {
-//     ydk::path::Repository repo{TEST_HOME};
-//     NetconfServiceProvider provider{repo, "127.0.0.1", "admin", "admin", 12022};
-//     ExecutorService es{};
-
-//     auto r_1 = make_unique<ydktest_sanity::Runner>();
-//     ydk::ietf_netconf::ValidateRpc rpc{};
-//     // rpc.source->config = // openconfg_bgp::Bgp -- create bgp object and assign
-//     bool reply = es.execute_rpc(provider, rpc);
-//     REQUIRE(reply);
```
Comment (abhikeshav) [inline sdk/cpp/tests/test_executor_service.cpp]: Remove commented lines

File: sdk/cpp/core/src/executor_service.cpp (hunk 15 above)
Comment (ylil93) [inline sdk/cpp/core/src/executor_service.cpp]: these lines don't exist anymore i think they're from an older commit

File: sdk/cpp/tests/test_executor_service.cpp (hunk 16 above)
Comment (ylil93) [inline sdk/cpp/tests/test_executor_service.cpp]: these lines don't exist anymore i think they're from an older commit
